### ISSUE TITLE
Leaderboard

### ISSUE THREAD

**Issue by niklasp** [Proof-Of-Chaos/website#8]
I think it would be nice to show some stats about owned NFT's on the NFT's page.

Some ideas, more are welcome:

    Total count (best distinct per rarity / ref I guess)
    Rarity spread (pie chart?)
    Rank of shelf (% owned compared to all owners)

For the last one @GregMcFly proposed to also show top shelves, which I feel is a great idea. This needs some more thought though. It might be cool to show them on the homepage, or perhaps even give it a page on its own.
The data can be fetched easily. We do need to decide on a scoring algorithm (weight of different refs / rarities and perhaps decorations). This part could also be moved to a separate new issue when initial NFT stats are done.

**Comment by gabrieljaegerde** [comment]
This issue already covers the leaderboard page quite nicely.

As our UX/UI lead maybe @niklasp you could propose some layouts for the page?

I think rendering the top shelves would be really cool but I am not sure about the complexity. 

This is an example app by the RMRK team for rendering RMRK2 nested NFTs: https://github.com/rmrk-team/rmrk2-examples/tree/master/projects/react-demo

Note that they use svg files for their collection. We are using png so the composer would need to be adjusted.

**Comment by gabrieljaegerde** [comment]
Doubt it. You have to compose it using the ipfs images. I can get more info for you but feel free to also ask directly in the [RMRK Implementers group](https://t.me/rmrkimpl) on telegram. Let me know if you prefer me to reach out.

**Comment by niklasp** [comment]
https://prerender.rmrk.link/1663411821945/11996468-3208723ec6f65df810-SHELF-VS-46.jpg

**Comment by gabrieljaegerde** [comment]
Nice. where did you find that?

**Comment by niklasp** [comment]
just the source code from singular. so i assume low quality images are generated when user changes resources of an nft

**Comment by niklasp** [comment]
so we just need a ranking (to be discussed in discord) a ui prototype and some endpoints getting:

- user ranking
- top n ranked shelves
- total shelves

**Comment by gabrieljaegerde** [comment]
The rank would need to be calculated directly in our app no?

**Comment by niklasp** [comment]
in my imagination it was part of the indexer? like a graphql endpoint to get all shelves but i am not deep enough into it. like on certain on chain events(new nft added to shelf, ...) update the rarity score of the shelf? whold that not work?

**Comment by gabrieljaegerde** [comment]
I think that should be all the way at the end of our indexer list and we should instead try to all do it on the frontend. I just don't know if it would be fast enough to do it in the browser?

Maybe you can look into Adam's script to find the X top shelves? The logic will be similar if not same.

**Comment by gabrieljaegerde** [comment]
Not sure how complete this is but looks like some work has been done: https://github.com/AlphabitTeam/RMRK-Indexer

**Comment by gabrieljaegerde** [comment]
I think that we should wait for rmrk3 with the leaderboard. It might not be worthwhile to invest in a rmrk2 indexer at this time anymore. I propose that we have a leaderboard in the form of a google sheet that automatically gets updated by Adam's script for now.

**Comment by gabrieljaegerde** [comment]
We now have the data available through a (centralized) endpoint. 

There are essentially 2 leaderboards: Main Leaderbaord and a temporary Adult Dragon Leaderboard

If a user's wallet is connected we should display their score and rank (maybe also shelf) at the top of the page for both leaderboards.

For the main leaderboard:
Below it we could have top 10 shelves with their prerendered images and scores as well (maybe also singular link).

For the adult dragon leaderboard we should display data on all the shelves that are provided by the endpoint but make it clear somehow that only the top 25 shelves will evolve to adult dragons.

**Comment by niklasp** [comment]
Cost estimation (80$ hourly rate)

    ui frame 5-6h (leaderboard tabs + 2 leaderboards (3h) + thumbnail preloader logic(0.5h) + queries(0.5h) + buffer(1-2h)

**Comment by gabrieljaegerde** [comment]
It would be great to show as well when the last update of the leaderboard was. Timestamp + block. @adamsteeber could you return that data too please.

**Comment by adamsteeber** [comment]
There is already a "block" value after the "score" array. I can try to figure out a way to get a Linux epoch timestamp in there but that might make more sense to do on the backend since you'd just have to pull the block number from the json and query the timestamp to display on the UI.

**Comment by adamsteeber** [comment]
Quote: $50/hr data processing and scripting
- 4 hours: scores algorithm development and scripting
- 2 hours: automation to backend

**Comment by niklasp** [comment]
block is enough we have the code to convert it already. thanks @adamsteeber

**Comment by gabrieljaegerde** [comment]
I think it would be v easy for @adamsteeber to return blocktime as well since blocks all have a timestamp. Might be better to do it like that otherwise we have the query the current block and then subtract.

**Comment by adamsteeber** [comment]
> I think it would be v easy for @adamsteeber to return blocktime as well since blocks all have a timestamp. Might be better to do it like that otherwise we have the query the current block and then subtract.

Blocktime is now included in the jsons under the `timestamp` key. It's an integer value of the linux epoch in milliseconds.

Quote: $35/hr novice js scripting
- 3 hours: write script to fetch timestamps directly from block numbers in jsons and overwrite the jsons with a timestamp value before scp to backend.

**Comment by gabrieljaegerde** [comment]
It sounds like you did an overkill. Could you share that script please.

**Comment by adamsteeber** [comment]
```
import { ApiPromise, WsProvider } from "@polkadot/api";
import fs from "fs";

async function main() {

  const shelfFile = fs.readFileSync('/root/gov/scores.json','utf8');
  const dragonFile = fs.readFileSync('/root/gov/adult_dragon_scores.json','utf8');
  const nftFile = fs.readFileSync('/root/gov/nftScores.json','utf8');
  const dragonNftFile = fs.readFileSync('/root/gov/adult_dragon_nftScores.json','utf8');

  const scores = JSON.parse(shelfFile);
  const dragonScores = JSON.parse(dragonFile);
  const nftScores = JSON.parse(nftFile);
  const dragonNftScores = JSON.parse(dragonNftFile);

  const blockNumberA = scores.block
  const blockNumberB = dragonScores.block
  const blockNumberC = nftScores.block
  const blockNumberD = dragonNftScores.block

  const wsProvider = new WsProvider('wss://kusama-rpc.rmrkdao.app');
  const api = await ApiPromise.create({ provider: wsProvider });
  const [hashA, hashB, hashC, hashD] = await Promise.all([
    api.rpc.chain.getBlockHash(blockNumberA),
    api.rpc.chain.getBlockHash(blockNumberB),
    api.rpc.chain.getBlockHash(blockNumberC),
    api.rpc.chain.getBlockHash(blockNumberD)
  ]);

  const [timeA, timeB, timeC, timeD] = await Promise.all([
    api.query.timestamp.now.at(hashA),
    api.query.timestamp.now.at(hashB),
    api.query.timestamp.now.at(hashC),
    api.query.timestamp.now.at(hashD)
  ]);

  scores.timestamp = parseInt(timeA.toString());
  dragonScores.timestamp = parseInt(timeB.toString());
  nftScores.timestamp = parseInt(timeC.toString());
  dragonNftScores.timestamp = parseInt(timeD.toString());

  fs.writeFileSync('/root/gov/scores.json',JSON.stringify(scores), {
    encoding: 'utf8',
    flag: 'w'
  });

  fs.writeFileSync('/root/gov/adult_dragon_scores.json',JSON.stringify(dragonScores), {
    encoding: 'utf8',
    flag: 'w'
  });

  fs.writeFileSync('/root/gov/nftScores.json',JSON.stringify(nftScores), {
    encoding: 'utf8',
    flag: 'w'
  });

  fs.writeFileSync('/root/gov/adult_dragon_nftScores.json',JSON.stringify(dragonNftScores), {
    encoding: 'utf8',
    flag: 'w'
  });
}

main()
  .catch(console.error)
  .finally(() => process.exit());
```

Most of the time was spent researching how to read and write files in node js, and figuring out where block timestamps are exposed. I wasn't able to find any documentation on `api.query.timestamp.now.at`. I also spent a good chunk of time integrating this script into the automation and testing to see if it was working properly.

Like I said, js is not my strong suit.

**Comment by gabrieljaegerde** [comment]
OK. Nice to see you getting familiar with this! I think this is similar to the logic we have in the front-end to convert blocks to timestamps then. It might be better to use the front end logic though as there we also have things like api reconnects in case of timeouts. I thought you would be able to retrieve the timestamp with ease by just querying it from the block directly and then saving it to the file with the block number together. I think I may have communicated this badly. That's my bad. 😬

@niklasp could you see if you can convert the block to a timestamp then in the front-end to show a last updated time.

**Comment by gabrieljaegerde** [comment]
Currently the images take very long to load on the leaderboards. Anyway to speed this up?

For the rank you seem to be taking the index and need to do a "+1"

**Comment by niklasp** [comment]
check if it is faster on vercel. you changed the image caching mode next has when migrating to ipns. i think we have no image cache atm with the next config we use.

underlying issue is the rmrk endpoint being slow.

**Comment by gabrieljaegerde** [comment]
Slow on vercel too. Can we implement a loading icon for the images?

Do you have a nice idea for how we could display the score boosts each respective NFT would yield?

**Comment by gabrieljaegerde** [comment]
Can we display all the shelves with score > 0 in the shelf leaderboard please.

**Comment by niklasp** [comment]
We cannot as long as we have no image caching, displaying the top 10 was a workaround for a lot of timeouts I head when displaying more. On vercel things work charming in the beta https://website-git-beta-proofofchaos.vercel.app/leaderboard/ (image cache activated here) but we cannot send 100+ requests to the rmrk prerenderer on each page load as it somehow breaks the response, it is probably throttled for external requests?

**Comment by gabrieljaegerde** [comment]
I don't mean images. Just score + wallet data. Right now we only have top 50.

**Comment by niklasp** [comment]
okay.

other topic, note to self, we can use this as image placeholder: https://github.com/vercel/next.js/blob/canary/examples/image-component/pages/shimmer.tsx

**Comment by niklasp** [comment]
closed in ce430d1d3
